jax/_src/numpy/lax_numpy.py

```diff
@@ -3847,6 +3847,36 @@ def diagflat(v, k=0):
   res = res.reshape(adj_length,adj_length)
   return res
 
+@_wraps(np.poly)
+def poly(seq_of_zeros):
+  seq_of_zeros = atleast_1d(seq_of_zeros)
+
+  dt = seq_of_zeros.dtype
+  seq_of_zeros = seq_of_zeros.astype(np.mintypecode(dt.char))
+
+  sh = seq_of_zeros.shape
+  if len(sh) == 2 and sh[0] == sh[1] != 0:
+    # import at runtime to avoid circular import
+    from . import linalg
+    seq_of_zeros = linalg.eigvals(seq_of_zeros)
+
+  if len(seq_of_zeros.shape) != 1:
+    raise ValueError("input must be 1d or non-empty square 2d array.")
+
+  if len(seq_of_zeros) == 0:
+    return 1.0
+  dt = seq_of_zeros.dtype
+  a = ones((1,), dtype=dt)
+  for k in range(len(seq_of_zeros)):
+    a = convolve(a, array([1, -seq_of_zeros[k]], dtype=dt), mode='full')
+
+  if issubclass(a.dtype.type, complexfloating):
+    roots = asarray(seq_of_zeros, complex)
+    roots_are_real = all(sort(roots) == sort(roots.conjugate()))
```

Comment:
I believe this check is not doing what you intend: the dtype of the returned value cannot depend on traced quantities, so the output will have a complex dtype regardless. I think this is probably a break from numpy's API that should be documented in the function's docstring via `lax_description` (search in the file to see how it's used)

Also, I think the logic of this check may be broken. e.g. if `roots=[1j, -1j]`, then `roots_are_real` is True.

Reply:
Upon investigating this further, I think we can remove this whole `if` block. 

The short explanation for this is that all this `if` block does is this: when all the imaginary parts are zero, throw away the imaginary parts of the answer by casting it to a real dtype.

I added this in my initial commit because I ported it over without looking at it in deeply. Since we can't change the type of an array based on its contents in JAX (please correct me if I'm wrong), we can't do what numpy does. Thus, we should just remove the whole block. 

I apologize for not finding this earlier. This was an oversight on my part. I should've figured out whether this was necessary when doing my first commit. I removed it in https://github.com/google/jax/commit/30633e3839cef7c7d29a72a90d1aeee7b8d9c701.

Here's the long explanation (feel free to ignore this as it's just for extra clarity):
<details>
<!---
- `np.poly` returns the coefficients of a polynomial whose zeros/roots are the inputs. 
- Let's call the returned result `c`. `c` represents the polynomial `c[0] * x**(N) + c[1] * x**(N-1) + ... + c[N-1] * x + c[N]`.
- Because `c[0]` is always 1 by [the definition of `np.poly`](https://numpy.org/doc/stable/reference/generated/numpy.poly.html), the polynomial can be written in the form: 
    - `(x-seq_of_zeros[0])*(x-seq_of_zeros[1])...`
    - Note how the coefficient of all the `x` terms is 1.
- `roots_are_real` tells us if `seq_of_zeros` are all complex conjugates.
- If `roots_are_real` is `True`, then the polynomial can be rewritten as:
    - `(x-seq_of_zeros[A])*(x-seq_of_zeros[A_CONJ])*(x-seq_of_zeros[B])*(x-seq_of_zeros[B_CONJ])*...`
    - where `seq_of_zeros[A]` and `seq_of_zeros[A_CONJ]` are conjugates (and similarly for `B` and all the rest).
- Since `(x-seq_of_zeros[A])*(x-seq_of_zeros[A_CONJ])` when expanded out will have all real coefficients (and similarly for `B` and all the rest), the whole polynomial will have all real coefficients. 
- Thus, `roots_are_real` just tells us whether or not the polynomial has all real coefficients. 
- In such cases, we can safely throw away all the imaginary parts of our answer and can safely cast our answer to have a real type.
-->
<ul>
<li><code>np.poly</code> returns the coefficients of a polynomial whose zeros/roots are the inputs. </li>
<li>Let&#39;s call the returned result <code>c</code>. <code>c</code> represents the polynomial <code>c[0] * x**(N) + c[1] * x**(N-1) + ... + c[N-1] * x + c[N]</code>.</li>
<li>Because <code>c[0]</code> is always 1 by <a href="https://numpy.org/doc/stable/reference/generated/numpy.poly.html">the definition of <code>np.poly</code></a>, the polynomial can be written in the form: <ul>
<li><code>(x-seq_of_zeros[0])*(x-seq_of_zeros[1])...</code></li>
<li>Note how the coefficient of all the <code>x</code> terms is 1.</li>
</ul>
</li>
<li><code>roots_are_real</code> tells us if <code>seq_of_zeros</code> are all complex conjugates.</li>
<li>If <code>roots_are_real</code> is <code>True</code>, then the polynomial can be rewritten as:<ul>
<li><code>(x-seq_of_zeros[A])*(x-seq_of_zeros[A_CONJ])*(x-seq_of_zeros[B])*(x-seq_of_zeros[B_CONJ])*...</code></li>
<li>where <code>seq_of_zeros[A]</code> and <code>seq_of_zeros[A_CONJ]</code> are conjugates (and similarly for <code>B</code> and all the rest).</li>
</ul>
</li>
<li>Since <code>(x-seq_of_zeros[A])*(x-seq_of_zeros[A_CONJ])</code> when expanded out will have all real coefficients (and similarly for <code>B</code> and all the rest), the whole polynomial will have all real coefficients. </li>
<li>Thus, <code>roots_are_real</code> just tells us whether or not the polynomial has all real coefficients. </li>
<li>In such cases, we can safely throw away all the imaginary parts of our answer and can safely cast our answer to have a real type.</li>
</ul>
</details>

Reply:
https://github.com/google/jax/commit/d3c3cfb43eeaa9bf30944b6550be0097ceefafd5 documents the differences with numpy's behavior via `lax_description`.